frontend/src/utils/account-send.ts

```diff
@@ -0,0 +1,179 @@
+    // This should never happen.
+    throw new Error(`Data to encrypt is not the correct length ${dataToEncrypt.length} instead of 64`);
+  }
+
+  return BigNumber.from(`0x${dataToEncrypt}`);
+};
+
+export const encryptAccountData = ({
+  address,
+  advancedMode,
+  usePublicKeyChecked,
+  encryptionCount,
+  viewingKey,
+  pubKey,
+}: EncryptAccountDataArgs) => {
+  const key = keccak256(toUtf8Bytes(`${viewingKey}${encryptionCount}`));
+  const data = buildAccountDataForEncryption({ address, advancedMode, usePublicKeyChecked, pubKey });
+  const encryptedData = data.xor(key);
+  return hexZeroPad(encryptedData.toHexString(), 32);
+};
+
+export const decryptData = ({ viewingKey, encryptionCount, encryptedAddress }: DecryptAccountDataArgs) => {
+  const key = keccak256(toUtf8Bytes(`${viewingKey}${encryptionCount}`));
+
+  const decryptedData = BigNumber.from(encryptedAddress).xor(key);
+  const hexData = decryptedData.toHexString();
+
+  const paritalPubKey = hexData.slice(44);
+  const advancedMode = hexData.slice(42, 43);
+  const usePublicKeyChecked = hexData.slice(43, 44);
+  return {
+    advancedMode,
+    usePublicKeyChecked,
+    address: hexData.slice(0, 42),
+    pubKey: paritalPubKey,
+  };
+};
+
+export const storeSend = async ({
+  recipientAddress,
+  chainId,
+  advancedMode,
+  usePublicKeyChecked,
+  viewingKey,
+  amount,
+  tokenAddress,
+  txHash,
+  userAddress,
+  provider,
+  pubKey,
+}: StoreSendArgs) => {
+  // Send history is scoped by chain
+  const key = `${LOCALFORAGE_ACCOUNT_SEND_KEY}-${userAddress}-${chainId}`;
+  const count =
+    ((await localforage.getItem(`${LOCALFORAGE_ACCOUNT_SEND_KEY}-count-${userAddress}-${chainId}`)) as number) || 0;
+  const checksummedRecipientAddress = await toAddress(recipientAddress, provider);
+  const encryptedData = encryptAccountData({
+    address: checksummedRecipientAddress,
+    advancedMode,
+    usePublicKeyChecked,
+    encryptionCount: count,
+    viewingKey,
+    pubKey,
+  });
+  const values = ((await localforage.getItem(key)) as AccountSendData[]) || [];
+  await localforage.setItem(key, [
+    ...values,
+    {
+      encryptedAddress: encryptedData,
+      amount,
+      tokenAddress,
+      dateSent: new Date(),
+      txHash,
+    },
+  ]);
+  await localforage.setItem(`${LOCALFORAGE_ACCOUNT_SEND_KEY}-count-${userAddress}-${chainId}`, count + 1);
+};
+
+export const fetchAccountSends = async ({ address, viewingKey, chainId }: FetchAccountSendArgs) => {
+  const key = `${LOCALFORAGE_ACCOUNT_SEND_KEY}-${address}-${chainId}`;
+  const values = ((await localforage.getItem(key)) as AccountSendDataWithEncryptedFields[]) || [];
+
+  const accountData = [] as AccountSendData[];
+  for (const [index, sendInfo] of values.entries()) {
+    const decryptedData = decryptData({
+      viewingKey,
+      encryptionCount: index,
+      encryptedAddress: sendInfo.encryptedAddress,
+    });
+    const recipientId = await lookupAddress(decryptedData.address, MAINNET_PROVIDER);
```

Comment:
`await` in a loop is going to be slow, we should pull this out and use `lookupOrReturnAddresses` to batch the lookups with multicall

Reply:
Handled here https://github.com/ScopeLift/umbra-protocol/pull/494/commits/a6e820e8cae2fa93f94dc277bcd10bf4789b9105

Reply:
Note: This does cause a regression because `lookupOrReturnAddresses` does not fallback to CNS

Reply:
@apbendi thoughts here? we never implemented batch CNS lookups in the same way. I'm ok not supporting CNS for now since we can always change this later since it just affects UX and doesn't actually affect encryption/decryption

Reply:
Now tracking in https://github.com/ScopeLift/umbra-protocol/issues/507, so marking this resolved